.github/workflows/onpullrequest-build-ubuntu.yml

```diff
@@ -1,19 +1,130 @@
+        path: depends/grpc
+    - name: Build grpc
+      run: if ! [ -d depends/grpc ] ; then scripts/install_grpc /usr v1.31.x depends/grpc ; fi
+
+  # Extract the commits of submodules for use by cache steps
+  onpr-submodules:
+    runs-on: ubuntu-20.04
+    outputs:
+      commits: ${{ steps.get-commits.outputs.commits }}
+    steps:
+    - uses: actions/checkout@v2
+    - name: Get Submodule Commits
+      id: get-commits
+      run: |
+        git submodule sync
+        echo "::set-output name=commits::"`git submodule status depends/libsodium | grep -oe '[0-9a-fA-F]\+' | head -c 8`-`git submodule status depends/libsnark | grep -oe '[0-9a-fA-F]\+' | head -c 8`
+
+  # Run the prover testing python scripts
+  prover-tests-linux:
+    runs-on: ubuntu-20.04
+    needs: [onpr-build-grpc, onpr-submodules]
+    strategy:
+      matrix:
+        curve: [ BLS12_377, ALT_BN128 ]
+    steps:
+    - uses: actions/checkout@v1
+      with:
+        submodules: recursive
+    - uses: actions/setup-node@v1
+      with:
+        node-version: 10
+    - name: Cache grpc
+      uses: actions/cache@v2
+      with:
+        key: grpc-1.31.x-${{ runner.os }}-test
+        path: depends/grpc
+    - name: Cache ccache
+      uses: actions/cache@v2
+      with:
+        key: prover-tests-ccache-${{ needs.onpr-submodules.outputs.commits }}-${{ matrix.curve }}-${{ runner.os }}-test
+        path: ~/.ccache
+    - name: Cache pip
+      uses: actions/cache@v2
+      with:
+        path: |
+          ~/.cache/pip
+          ~/Library/Caches/pip
+        key: prover-tests-pip-${{ hashFiles('**/setup.py') }}-${{ runner.os }}
+    - name: Cache npm
+      uses: actions/cache@v2
+      with:
+        path: |
+          ~/.npm
+          depends/ganache-cli/node_modules
+        key: prover-tests-npm-${{ hashFiles('**/package-lock.json') }}-${{ runner.os }}
+    - name: Install dependencies
+      run: |
+        scripts/install_grpc /usr v1.31.x depends/grpc
+        sudo apt install -y ccache
+    - name: Execute
+      run: CI_CONFIG=Release CI_CURVE=${{ matrix.curve }} CI_PROVER_TESTS=1 scripts/ci build
+
+  # Run all unit tests and integration tests
+  integration-tests-linux:
+    runs-on: ubuntu-20.04
+    needs: [onpr-build-grpc, onpr-submodules]
     strategy:
       matrix:
         curve: [ BLS12_377, ALT_BN128 ]
     steps:
     - uses: actions/checkout@v1
       with:
         submodules: recursive
+    - uses: actions/setup-node@v1
+      with:
+        node-version: 10
+    - name: Cache grpc
+      uses: actions/cache@v2
+      with:
+        key: grpc-1.31.x-${{ runner.os }}-test
+        path: depends/grpc
+    # ccache in this job seems to make the tests crash occasionally. Disabling
+    # until the cause is understood.
+    # - name: Cache ccache
+    #   uses: actions/cache@v2
+    #   with:
+    #     key: integration-tests-ccache-${{ needs.onpr-submodules.outputs.commits }}-${{ matrix.curve }}-${{ runner.os }}-test
+    #     path: ~/.ccache
```

Comment:
TBH, I'm not sure.  It may be possible (assuming there is synchronisation between the two jobs so they don't both try to populate the cache at the same time).  When I started seeing crashes which seemed to be due to ccache on linux, I think the one of the first things I tried was to separate the caches. I still got a crash in some situations, hence this is all disabled for now.  If we get it working, we can try combining them.

Reply:
> I still got a crash in some situations, hence this is all disabled for now. If we get it working, we can try combining them.

Got it.

> When I started seeing crashes which seemed to be due to ccache on linux

Out of curiosity, do you have the trace of the ccache error you had?

Reply:
> Out of curiosity, do you have the trace of the ccache error you had?

I don't have one immediately available, but it was not particularly helpful.  The unit tests crashed, so all output was hidden by gtest. gtest just reported `Illegal ...` (i.e. a truncated description).

My guess was that this was "illegal instruction" (which i could imagine happening for a few different reasons, although a quick search did not show up anything about ccache on github actions).  But to get any definitive information would require more investigation.

Reply:
Got it, thanks.
Let's take it incrementally and get this one merged for now. It extends the CI in a meaningful way. I'll open a ticket for this issue to stay on our radar.

Reply:
See: https://github.com/clearmatics/zeth/issues/351